### PR TITLE
Implement random.choice

### PR DESCRIPTION
**Pull request by KotaroSetoyama**

### CODE REVIEW

**Comment by delta2323** [comment]
Please add the docstring to the document.

File: cupy/random/sample.py [hunk 1]
```diff
@@ -136,4 +139,52 @@ def random_sample(size=None, dtype=float):
     return rs.random_sample(size=size, dtype=dtype)
 
 
-# TODO(okuta): Implement choice
+def choice(a, size=None, replace=True, p=None):
+    """Return an array of random values from a given 1-D array.
+
+    Each element of returned values are independently sampled
```
Comment (delta2323) [inline cupy/random/sample.py]: Suggestion of the wording: Each element of the returned array ...

Comment (delta2323) [inline cupy/random/sample.py]: are -> is

File: cupy/random/sample.py [hunk 2]
```diff
@@ -136,4 +139,52 @@ def random_sample(size=None, dtype=float):
...
+    from a according to p or uniformly.
+
+    Args:
+        a (1-D array-like or int):
+            If an ndarray, a random sample is generated from its elements.
```
Comment (delta2323) [inline cupy/random/sample.py]: Which `ndarray`? NumPy or CuPy?

File: cupy/random/sample.py [hunk 3]
```diff
@@ -136,4 +139,52 @@ def random_sample(size=None, dtype=float):
...
+            pop_size = operator.index(a.item())
+        except TypeError:
+            raise ValueError("a must be 1-dimensional or an integer")
+        if pop_size <= 0:
+            raise ValueError("a must be greater than 0")
+    elif a.ndim != 1:
+        raise ValueError("a must be 1-dimensional")
+    else:
+        pop_size = a.shape[0]
+        if pop_size is 0:
```
Comment (delta2323) [inline cupy/random/sample.py]: Is there any reason why we use `is` for comparing integers? If not, `pop_size == 0` is better.

File: cupy/random/sample.py [hunk 4]
```diff
@@ -136,4 +139,52 @@ def random_sample(size=None, dtype=float):
...
+
+    """
+    a = cupy.array(a, copy=False)
+    if a.ndim == 0:
+        try:
+            pop_size = operator.index(a.item())
+        except TypeError:
+            raise ValueError("a must be 1-dimensional or an integer")
+        if pop_size <= 0:
+            raise ValueError("a must be greater than 0")
```
Comment (delta2323) [inline cupy/random/sample.py]: ditto (use single quotation)

File: cupy/random/sample.py [hunk 5]
```diff
@@ -136,4 +139,52 @@ def random_sample(size=None, dtype=float):
...
+    elif a.ndim != 1:
+        raise ValueError("a must be 1-dimensional")
```
Comment (delta2323) [inline cupy/random/sample.py]: ditto (use single quotation)

File: cupy/random/sample.py [hunk 6]
```diff
@@ -136,4 +139,52 @@ def random_sample(size=None, dtype=float):
...
+    else:
+        pop_size = a.shape[0]
+        if pop_size is 0:
+            raise ValueError("a must be non-empty")
```
Comment (delta2323) [inline cupy/random/sample.py]: ditto (use single quotation)

File: cupy/random/sample.py [hunk 7]
```diff
@@ -136,4 +139,52 @@ def random_sample(size=None, dtype=float):
...
+
+    if p is not None:
+        p = cupy.array(p)
+        if p.ndim != 1:
+            raise ValueError("p must be 1-dimensional")
```
Comment (delta2323) [inline cupy/random/sample.py]: ditto (use single quotation)

File: cupy/random/sample.py [hunk 8]
```diff
@@ -136,4 +139,52 @@ def random_sample(size=None, dtype=float):
...
+        if len(p) != pop_size:
+            raise ValueError("a and p must have same size")
```
Comment (delta2323) [inline cupy/random/sample.py]: ditto (use single quotation)

File: cupy/random/generator.py [hunk 9]
```diff
@@ -235,6 +235,71 @@ def uniform(self, low=0.0, high=1.0, size=None, dtype=float):
...
+        shape = size
+        size = numpy.prod(shape)
+
+        if p is not None:
+            p = cupy.broadcast_to(p, (size, a_size))
+            index = cupy.argmax(cupy.log(p) -
+                                cupy.random.gumbel(size=(size, a_size)),
+                                axis=1)
+            if type(shape) != int:
+                index.shape = shape
```
Comment (delta2323) [inline cupy/random/generator.py]: It seems odd to me because this line try to change the shape of `CuPy` array. Could you explain the intent of this line?

File: cupy/random/generator.py [hunk 10]
```diff
@@ -235,6 +235,71 @@ def uniform(self, low=0.0, high=1.0, size=None, dtype=float):
...
+        else:
+            index = cupy.random.randint(0, a_size, size=shape)
+
+        if a.ndim == 0:
+            return index
+
+        if index.ndim == 0:
+            # size = ()
```
Comment (delta2323) [inline cupy/random/generator.py]: Remove this comment.

File: cupy/random/generator.py [hunk 11]
```diff
@@ -235,6 +235,71 @@ def uniform(self, low=0.0, high=1.0, size=None, dtype=float):
...
+            raise NotImplementedError
+        shape = size
+        size = numpy.prod(shape)
+
+        if p is not None:
+            p = cupy.broadcast_to(p, (size, a_size))
+            index = cupy.argmax(cupy.log(p) -
+                                cupy.random.gumbel(size=(size, a_size)),
+                                axis=1)
+            if type(shape) != int:
```
Comment (delta2323) [inline cupy/random/generator.py]: Use `six.interger_types`

File: tests/cupy_tests/random_tests/test_generator.py [hunk 12]
```diff
@@ -253,6 +253,88 @@ def test_goodness_of_fit_2(self):
...
+@testing.gpu
+class TestChoice(unittest.TestCase):
+
+    def setUp(self):
+        self.rs = generator.RandomState()
+
+    @condition.repeat(10)
+    def test_within_choice_and_shape(self):
+        val = self.rs.choice(a=self.a, size=self.size, p=self.p).get()
+        minimum = -1 if type(self.a) == int else 0
```
Comment (delta2323) [inline tests/cupy_tests/random_tests/test_generator.py]: ditto (use `six.interger_types`)

File: tests/cupy_tests/random_tests/test_generator.py [hunk 13]
```diff
@@ -253,6 +253,88 @@ def test_goodness_of_fit_2(self):
...
+        maximum = 3 if type(self.a) == int else 4
```
Comment (delta2323) [inline tests/cupy_tests/random_tests/test_generator.py]: ditto (use `six.interger_types`)

File: tests/cupy_tests/random_tests/test_generator.py [hunk 14]
```diff
@@ -253,6 +253,88 @@ def test_goodness_of_fit_2(self):
...
+        lower = 0 if type(self.a) == int else 1
+        self.assertEqual(lower, val)
+
+    @condition.retry(20)
+    def test_upper_bound(self):
+        val = self.rs.choice(a=self.a, size=self.size, p=self.p).get()
+        val = val.item() if self.size == () else val.item(0)
+        upper = 2 if type(self.a) == int else 3
+        self.assertEqual(upper, val)
+
```
Comment (delta2323) [inline tests/cupy_tests/random_tests/test_generator.py]: Could you check the shape of the returned array of `choice` is expected one?

**Comment by delta2323** [comment]
Could you remove [WIP] when you finish implementation?

File: tests/cupy_tests/random_tests/test_generator.py [hunk 15]
```diff
@@ -253,6 +253,96 @@ def test_goodness_of_fit_2(self):
...
+@testing.parameterize(
+    {'a': 3.1, 'size': 1, 'p': [0.1, 0.1, 0.8]},
+    {'a': None, 'size': 1, 'p': [0.1, 0.1, 0.8]},
+    {'a': -3, 'size': 1, 'p': [0.1, 0.1, 0.8]},
+    {'a': [[0, 1], [2]], 'size': 1, 'p': [0.1, 0.1, 0.8]},
+    {'a': 3, 'size': 1, 'p': [[0.1, 0.1], [0.8]]},
+    {'a': 2, 'size': 1, 'p': [0.1, 0.1, 0.8]},
+    {'a': 3, 'size': 1, 'p': [-0.1, 0.3, 0.8]},
+    {'a': 3, 'size': 1, 'p': [0.1, 0.1, 0.7]},
```
Comment (delta2323) [inline tests/cupy_tests/random_tests/test_generator.py]: Could you add a test case where `a` is an empty array for checking if [this error](https://github.com/pfnet/chainer/pull/1979/files?diff=unified#diff-bdf8bf88a45064e9d8d316ebd0d0f6fbR262) is raised appropriately.

File: cupy/random/generator.py [hunk 16]
```diff
@@ -235,6 +235,74 @@ def uniform(self, low=0.0, high=1.0, size=None, dtype=float):
...
+        if isinstance(a, six.integer_types):
+            a_size = a
+            if a_size <= 0:
+                raise ValueError('a must be greater than 0')
+            a = cupy.array(a, copy=False)
+        else:
+            a = cupy.array(a, copy=False)
+            if a.ndim != 1:
+                raise ValueError('a must be 1-dimensional')
+            elif a.dtype != 'int':
```
Comment (delta2323) [inline cupy/random/generator.py]: It seems to me that when `a` is array_like, its dtype need not to be integer. Actually, a is an array of strings in some example of `NumPy`.

File: cupy/random/generator.py [hunk 17]
```diff
@@ -235,6 +235,74 @@ def uniform(self, low=0.0, high=1.0, size=None, dtype=float):
...
+            raise NotImplementedError
+        shape = size
+        size = numpy.prod(shape)
+
+        if p is not None:
+            p = cupy.broadcast_to(p, (size, a_size))
+            index = cupy.argmax(cupy.log(p) -
+                                cupy.random.gumbel(size=(size, a_size)),
+                                axis=1)
+            index = index.astype(cupy.int32, copy=False)
```
Comment (delta2323) [inline cupy/random/generator.py]: Could you add a comment why we need this line (e.g. to align the dtype with NumPy).

File: cupy/random/generator.py [hunk 18]
```diff
@@ -235,6 +235,74 @@ def uniform(self, low=0.0, high=1.0, size=None, dtype=float):
...
+                index = cupy.reshape(index, shape)
+        else:
+            index = cupy.random.randint(0, a_size, size=shape)
+
+        if a.ndim == 0:
+            return index
+
+        if index.ndim == 0:
+            res = cupy.empty((), dtype=a.dtype)
+            res[()] = a[index]
```
Comment (delta2323) [inline cupy/random/generator.py]: You can write as `res = cupy.array(a[index], dtype=a.dtype)` instead.

File: tests/cupy_tests/random_tests/test_generator.py [hunk 19]
```diff
@@ -253,6 +253,96 @@ def test_goodness_of_fit_2(self):
         self.assertTrue(hypothesis.chi_square_test(counts, expected))
 
 
+@testing.parameterize(
+    {'a': 3, 'size': 2, 'p': None},
+    {'a': 3, 'size': 2, 'p': [0.3, 0.3, 0.4]},
+    {'a': 3, 'size': (5, 5), 'p': [0.3, 0.3, 0.4]},
+    {'a': 3, 'size': (), 'p': None},
+    {'a': [1, 2, 3], 'size': 2, 'p': [0.3, 0.3, 0.4]},
```
Comment (delta2323) [inline tests/cupy_tests/random_tests/test_generator.py]: Could you add a test case where `a` is an one-dimentional array whose `dtype` is different from integer.

File: tests/cupy_tests/random_tests/test_generator.py [hunk 20]
```diff
@@ -253,6 +253,96 @@ def test_goodness_of_fit_2(self):
...
+)
+@testing.gpu
+class TestChoice(unittest.TestCase):
+
+    def setUp(self):
+        self.rs = generator.RandomState()
+
+    def test_shape(self):
```
Comment (delta2323) [inline tests/cupy_tests/random_tests/test_generator.py]: Could you rename the test name because this test case actually checks both `dtype` and `shape`.

File: tests/cupy_tests/random_tests/test_generator.py (hunk 19 above)
Comment (delta2323) [inline tests/cupy_tests/random_tests/test_generator.py]: Do you have special reason why you chose `[1, 2, 3]` as the value of `a`? If nothing, how about changing the value of `a` in the last test case from `[1, 2, 3]` to `[0, 1, 2]`? It will remove cumbersome branchings in e.g. [here](https://github.com/pfnet/chainer/pull/1979/files?diff=unified#diff-d5961550728ebd08cfb44c0e5074cd97R281).

**Comment by KotaroSetoyama** [comment]
I confirmed NumPy's return dtype is `int64` or `float64`, and accommodated the CuPy's return dtype to them. And because there is no case to input `string` values as an argument `a`, I have not implemented such case.

**Comment by delta2323** [comment]
@KotaroSetoyama The [document of `numpy.random.choice`](https://docs.scipy.org/doc/numpy-dev/reference/generated/numpy.random.choice.html) has an example of `a` being a list of strings. That is why I ask you if we need string test case. But I come to think we can do it afterwards and it is more important to finish the initial implementation. So it is OK to skip the case.

**Comment by delta2323** [comment]
LGTM. I am running Jenkins to see all tests pass. After checking that, I will merge this PR. Thank you.

**Comment by delta2323** [comment]
Could you merge the latest master branch as Jenkins failed at unit tests that do not seem to be relevant to this PR.

**Comment by delta2323** [comment]
Just a memorandum. As I have shared with @KotaroSetoyama, Jenkins failed with the following message. We are investing this problem, but could not find out the reason.

```
======================================================================
13:17:53 ERROR: test_concat_arrays_gpu (test_convert.TestConcatExamples)
13:17:53 ----------------------------------------------------------------------
13:17:53 Traceback (most recent call last):
13:17:53   File "/work/chainer/tests/chainer_tests/dataset_tests/test_convert.py", line 36, in test_concat_arrays_gpu
13:17:53     arrays = self.get_arrays_to_concat(cuda.cupy)
13:17:53   File "/work/chainer/tests/chainer_tests/dataset_tests/test_convert.py", line 14, in get_arrays_to_concat
13:17:53     return [xp.random.rand(2, 3) for _ in range(5)]
13:17:53   File "/work/chainer/cupy/random/sample.py", line 28, in rand
13:17:53     return random_sample(size=size, dtype=dtype)
13:17:53   File "/work/chainer/cupy/random/sample.py", line 135, in random_sample
13:17:53     rs = generator.get_random_state()
13:17:53   File "/work/chainer/cupy/random/generator.py", line 348, in get_random_state
13:17:53     rs = RandomState(os.getenv('CHAINER_SEED'))
13:17:53   File "/work/chainer/cupy/random/generator.py", line 45, in __init__
13:17:53     self._generator = curand.createGenerator(method)
13:17:53   File "cupy/cuda/curand.pyx", line 92, in cupy.cuda.curand.createGenerator (cupy/cuda/curand.cpp:1436)
13:17:53     cpdef size_t createGenerator(int rng_type) except *:
13:17:53   File "cupy/cuda/curand.pyx", line 96, in cupy.cuda.curand.createGenerator (cupy/cuda/curand.cpp:1374)
13:17:53     check_status(status)
13:17:53   File "cupy/cuda/curand.pyx", line 85, in cupy.cuda.curand.check_status (cupy/cuda/curand.cpp:1226)
13:17:53     raise CURANDError(status)
13:17:53 CURANDError: CURAND_STATUS_INITIALIZATION_FAILED
```

**Comment by unnonouno** [comment]
I also tried to check the error, and faced the same error.

**Comment by unnonouno** [comment]
I found that it cause no error when multiprocess is disabled.

File: tests/cupy_tests/random_tests/test_generator.py [hunk 21]
```diff
@@ -253,6 +253,100 @@ def test_goodness_of_fit_2(self):
         self.assertTrue(hypothesis.chi_square_test(counts, expected))
 
 
+@testing.parameterize(
+    {'a': 3, 'size': 2, 'p': None},
+    {'a': 3, 'size': 2, 'p': [0.3, 0.3, 0.4]},
+    {'a': 3, 'size': (5, 5), 'p': [0.3, 0.3, 0.4]},
+    {'a': 3, 'size': (5, 5), 'p': cupy.array([0.3, 0.3, 0.4])},
```
Comment (okuta) [inline tests/cupy_tests/random_tests/test_generator.py]: Please use `numpy.array`.
Do not use `cupy.array` in decorator. Because, this line runs in cpu environment.

File: tests/cupy_tests/random_tests/test_generator.py [hunk 22]
```diff
@@ -253,6 +253,100 @@ def test_goodness_of_fit_2(self):
...
+    {'a': 3, 'size': (), 'p': None},
+    {'a': [0, 1, 2], 'size': 2, 'p': [0.3, 0.3, 0.4]},
+    {'a': cupy.array([0.0, 1.0, 2.0]), 'size': 2, 'p': [0.3, 0.3, 0.4]},
```
Comment (okuta) [inline tests/cupy_tests/random_tests/test_generator.py]: ditto

**Comment by delta2323** [comment]
> I found that it cause no error when multiprocess is disabled.

I see. Then, should we run the test locally to confirm this PR works without fails?

**Comment by unnonouno** [comment]
It passed the test. @delta2323 Can i merge this PR?

**Comment by delta2323** [comment]
@unnonouno ok. go ahead
